sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala

```diff
@@ -43,25 +43,29 @@ object GenerateUnsafeProjection extends CodeGenerator[Seq[Expression], UnsafePro
     case _ => false
   }
 
-  // TODO: if the nullability of field is correct, we can use it to save null check.
   private def writeStructToBuffer(
       ctx: CodegenContext,
       input: String,
       index: String,
-      fieldTypes: Seq[DataType],
+      fieldTypeAndNullables: Seq[(DataType, Boolean)],
```

Comment:
shall we create a class for `(DataType, Boolean)`? it can also be used in https://github.com/apache/spark/pull/22063

Reply:
I think that it would be good since it is used at `JavaTypeInference` and `higherOrderFunctions`.
cc @ueshin

Reply:
@cloud-fan What name of the case class do you suggest?  `DataTypeNullable`, or others?

Reply:
@cloud-fan I found this case class `case class Schema(dataType: DataType, nullable: Boolean)` at two places.
1. `ScalaReflection.Schema`
1. `SchemaConverters.SchemaType`

Do we use one of them? Or do we define `org.apache.spark.sql.types.Schema`?

Reply:
I think you can define one within this class for readability. Probably we should take a look to deduplicate them for the instances you pointed out, but not sure yet for now if how much that deduplication improve the readability and if that's worth. Those are all look rather defined and used within small scopes rather locally.